includes/backup.js

```diff
@@ -13,267 +13,100 @@
+const { createWriteStream } = require('node:fs');
+const { pipeline } = require('node:stream/promises');
+const { Backup } = require('./backupMappings.js');
+const { BackupError, HTTPError, convertResponseError } = require('./error.js');
+const logFileSummary = require('./logfilesummary.js');
+const logFileGetBatches = require('./logfilegetbatches.js');
 const spoolchanges = require('./spoolchanges.js');
-const logfilesummary = require('./logfilesummary.js');
-const logfilegetbatches = require('./logfilegetbatches.js');
-
-/**
- * Read documents from a database to be backed up.
- *
- * @param {string} db - `@cloudant/cloudant` DB object for source database.
- * @param {number} blocksize - number of documents to download in single request
- * @param {number} parallelism - number of concurrent downloads
- * @param {string} log - path to log file to use
- * @param {boolean} resume - whether to resume from an existing log file
- * @returns EventEmitter with following events:
- *  - `received` - called with a block of documents to write to backup
- *  - `error` - on error
- *  - `finished` - when backup process is finished (either complete or errored)
- */
-module.exports = function(db, options) {
-  const ee = new events.EventEmitter();
-  const start = new Date().getTime(); // backup start time
-  const batchesPerDownloadSession = 50; // max batches to read from log file for download at a time (prevent OOM)
-
-  function proceedWithBackup() {
-    if (options.resume) {
-      // pick up from existing log file from previous run
-      downloadRemainingBatches(options.log, db, ee, start, batchesPerDownloadSession, options.parallelism);
-    } else {
-      // create new log file and process
-      spoolchanges(db, options.log, options.bufferSize).then(() => {
-        downloadRemainingBatches(options.log, db, ee, start, batchesPerDownloadSession, options.parallelism);
-      }).catch((err) => {
-        ee.emit('error', err);
-      });
-    }
-  }
-
-  validateBulkGetSupport(db, function(err) {
-    if (err) {
-      return ee.emit('error', err);
-    } else {
-      proceedWithBackup();
-    }
-  });
-
-  return ee;
-};
+const { MappingStream, WritableWithPassThrough, DelegateWritable, SideEffect } = require('./transforms.js');
 
 /**
  * Validate /_bulk_get support for a specified database.
  *
- * @param {string} db - nodejs-cloudant db
- * @param {function} callback - called on completion with signature (err)
+ * @param {object} db - object for connection to source database containing name, service and url
  */
-function validateBulkGetSupport(db, callback) {
-  db.service.postBulkGet({ db: db.db, docs: [] }).then(() => { callback(); }).catch(err => {
-    err = error.convertResponseError(err, function(err) {
-      switch (err.status) {
-        case undefined:
+async function validateBulkGetSupport(db) {
+  return db.service.postBulkGet({ db: db.db, docs: [] })
+    .catch(err => {
+      err = convertResponseError(err, function(err) {
```

Comment:
So we throw from inside error factory and doing not working reassign on `err` which we doesn't return anyway here? This looks wrong, it seems that instead the error factory should return composed error and catch should return or re-throw it as `throw convertResponseError` instead.  Honestly this would be less confusing as `try {} catch` around `postBulkGet` instead.